src/music/MusicManager.ts

```diff
@@ -0,0 +1,185 @@
+        [ChordFunction.SUBDOMINANT, ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.AMBIGUOUS, ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT, ChordFunction.AMBIGUOUS]
+    ];
+
+    const FunctionDegrees: Record<ChordFunction, NonEmptyArray<number>> = {
+        "tonic": [0, 5], // I and VI
+        "subdominant": [1, 3], // II and IV
+        "dominant": [4, 6], // V and VII
+        "ambiguous": [2] // III
+    };
+
+    function generateChordProgression(scale: Scale, root: number, extensions: number = 4): number[][] {
+        const pitchClass: number[] = Scales.getPitchClass(scale);
+        const backHalf: ChordFunction[] = Random.fromArray(backHalves);
+        // always start from the tonic and add a valid back half
+        // this way it'll always fit the tonic -> subdominant -> dominant pattern
+        const chordDegrees: number[] = [0].concat(backHalf.map(f => Random.fromArray(FunctionDegrees[f])));
+        return chordDegrees.map(degree => {
+            const chord = [];
+            for (let i = 0; i < extensions; i++) {
+                // step up scale in thirds starting from chord's root
+                chord[i] = Scales.indexIntoPitchClass(pitchClass, degree + 2 * i) + root;
+            }
+            return chord;
+        });
+    }
+
+    function generateDrumLoop(): Drums[] {
+        const drumLoop: Drums[] = [];
+        const subdivision: number[] = state.rhythm.subdivision;
+        let onKick: boolean = true;
+        for (let i = 0; i < subdivision.length; i++) {
+            let acc: number = subdivision[i] - 1;
+            if (onKick) {
+                drumLoop.push(Drums.KICK);
+            } else {
+                drumLoop.push(Drums.SNARE);
+            }
+            onKick = !onKick;
+            while (acc > 0) {
+                if (acc === 1 && i === subdivision.length - 1) {
+                    drumLoop.push(Drums.OPEN_HI_HAT);
+                } else {
+                    drumLoop.push(Drums.CLOSED_HI_HAT);
+                }
+                acc--;
+            }
+        }
+        return drumLoop;
+    }
+
+    function scheduleDrum(start: number, drum: Drums): void {
+        const drumOut: AudioNode = drumkit.scheduleHit(context, start, drum);
+        drumOut.connect(masterGain);
+    }
+
+    function scheduleNote(note: Note, inst: Instrument): void {
+        const instOut: AudioNode = inst.scheduleNote(context, note);
+        instOut.connect(masterGain);
+    }
+
+    // shift note through the octaves until it ends up between min and max
+    function constrainNote(note: number, min: number, max: number): number {
+        while (note < min) {
+            note += 12;
+        }
+        while (note > max) {
+            note -= 12;
+        }
+        return note;
```

Comment:
I'm pretty sure this can be replaced by something like
```typescript
if (note < min) {
    note += 12 * Math.ceil((min - note) / 12);
}
if (note > max) {
    note -= 12 * Math.ceil((note - max) / 12);
}
```
This can definitely be done without looping like this.

Reply:
Also, you don't actually get the guarantee that `note` is between `min` and `max` inclusive if your range is smaller than 12. You should mention this in the docs.